peer.go

```diff
@@ -938,137 +966,186 @@ func (p *peer) handleLocalClose(req *htlcswitch.ChanClose) {
 			"unknown", chanID)
 		peerLog.Errorf(err.Error())
-		req.Err <- err
-		return
+		if localReq != nil {
+			localReq.Err <- err
+		}
+		return nil, 0
 	}
 
-	// Calculate a fee rate that we believe to be fair and will ensure a
-	// timely confirmation.
-	//
-	// TODO(bvu): with a dynamic fee implementation, we will compare this
-	// to the fee proposed by the responder in their ClosingSigned message.
-	feeRate := p.server.cc.feeEstimator.EstimateFeePerWeight(1) * 1000
-
-	// We agree with the proposed channel close transaction and fee rate,
-	// so generate our signature.
-	initiatorSig, proposedFee, err := channel.CreateCloseProposal(feeRate)
-	if err != nil {
-		req.Err <- err
-		return
-	}
-	initSig := append(initiatorSig, byte(txscript.SigHashAll))
-
-	// Complete coop close transaction with the signatures of the close
-	// initiator and responder.
-	responderSig := msg.Signature
-	respSig := append(responderSig.Serialize(), byte(txscript.SigHashAll))
-	closeTx, err := channel.CompleteCooperativeClose(initSig, respSig,
-		feeRate)
+	// We now consider the fee proposed by the peer, together with the fee we
+	// last proposed (if any). This method will in case more fee negotiation is
+	// necessary send a new ClosingSigned message to the peer with our new
+	// proposed fee. In case we can agree on a fee, it will assemble the close
+	// transaction, and we can go on to broadcasting it.
+	closeTx, ourSig, ourFee, err := p.negotiateFeeAndCreateCloseTx(channel, msg,
+		lastSig, lastFee)
 	if err != nil {
-		req.Err <- err
-		// TODO(roasbeef): send ErrorGeneric to other side
-		return
+		if localReq != nil {
+			localReq.Err <- err
+		}
+		return nil, 0
 	}
 
-	// As we're the initiator of this channel shutdown procedure we'll now
-	// create a mirrored close signed message with our completed signature.
-	parsedSig, err := btcec.ParseSignature(initSig, btcec.S256())
-	if err != nil {
-		req.Err <- err
-		return
+	// If closeTx == nil it means that we did not agree on a fee, but we proposed
+	// a new fee to the peer. Return the signature used for this new proposal,
+	// and the fee we proposed, for use when we get a reponse.
+	if closeTx == nil {
+		return ourSig, ourFee
 	}
-	closingSigned := lnwire.NewClosingSigned(chanID, proposedFee, parsedSig)
-	p.queueMsg(closingSigned, nil)
 
-	// Finally, broadcast the closure transaction to the network.
+	// We agreed on a fee, and we can broadcast the closure transaction to the
+	// network.
 	peerLog.Infof("Broadcasting cooperative close tx: %v",
 		newLogClosure(func() string {
 			return spew.Sdump(closeTx)
 		}))
+
+	chanPoint := channel.ChannelPoint()
 	if err := p.server.cc.wallet.PublishTransaction(closeTx); err != nil {
-		peerLog.Errorf("channel close tx from "+
-			"ChannelPoint(%v) rejected: %v",
-			req.ChanPoint, err)
-		// TODO(roasbeef): send ErrorGeneric to other side
-		return
+		if strings.Contains(err.Error(), "already exists") ||
```

Comment:
Agreed. Should I create a PR to your fork of `btcwallet` (and neutrino)? Also, should the possible errors be specified in the `WalletController` interface?